### PR TITLE
Fix snapshot deletion on template creation failure

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
### Description

This PR addresses an issue as part of #8946 

The issue I've observed is not while taking the snapshot but while creating the template from the snapshot (stack trace also refers the same).

I could not reproduce the original issue of failed snapshot showing as backedup state rather than error (it might have already fixed after 4.17.2), but I saw another serious issue.

The issue is whenever a snapshot is used to create a template or volume and if there is failure in backing up the snapshot to the secondary store and as part of handling that failure MS is deleting the snapshot in primary storage itself.

These changes are introduced as part of the PR https://github.com/apache/cloudstack/pull/5297

1. Create a snapshot of a volume (set snapshot.backup.to.secondary = False)
2. Create a template from that snapshot
3. As part of the creation, MS first tries to backup the snapshot to the secondary storage
4. I've made it fail
5. MS recognized the failure and as part of failure it is deleting the snapshot on the primary storage (also marking the snapshot_store_ref entry for primary store role as "Destroyed")

Here in this PR, the fix is to handle this case of not deleting the snapshot on primary storage.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

First lets check the successful scenario

1. Create a snapshot of a volume (set snapshot.backup.to.secondary = False)
2. Create a template from that snapshot
3. As part of the creation, MS first tries to backup the snapshot to the secondary storage
4. One new entry in the snapshot_store_ref table will be seen for store_role "Image" in "creating" state
5. After the successful creation of template, this entry will be deleted.

Now the failure scenario

1. Create a snapshot of a volume (set snapshot.backup.to.secondary = False)
2. Create a template from that snapshot
3. As part of the creation, MS first tries to backup the snapshot to the secondary storage
4. One new entry in the snapshot_store_ref table will be seen for store_role "Image" in "creating" state
5. Make the backup operation fail (I've played with the debugger to fail the operation)
6. Observe the snapshot_store_ref table, this entry will be deleted and also keeps the existing entry for store_role "Primary" as "Ready" state. (Previously this row was marked as "Destroyed")

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

File: server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java [hunk 1]
```diff
@@ -101,15 +101,21 @@ public void expungeTemporarySnapshot(boolean kvmSnapshotOnlyInPrimaryStorage, Sn
...
 
-        logger.debug(String.format("Expunging snapshot [%s] due to it is a temporary backup to create a volume from snapshot. It is occurring because the global setting [%s]"
-          + " has the value [%s].", snapInfo.getId(), SnapshotInfo.BackupSnapshotAfterTakingSnapshot.key(), backupSnapshotAfterTakingSnapshot));
-
-        try {
-            snapshotService.deleteSnapshot(snapInfo);
-        } catch (CloudRuntimeException ex) {
-            logger.warn(String.format("Unable to delete the temporary snapshot [%s] on secondary storage due to [%s]. We still will expunge the database reference, consider"
-              + " manually deleting the file [%s].", snapInfo.getId(), ex.getMessage(), snapInfo.getPath()), ex);
+        if (!snapInfo.getDataStore().getRole().equals(DataStoreRole.Image)) {
```
Comment (DaanHoogland) [inline server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java:104]:
```suggestion
        if (!DataStoreRole.Image.equals(snapInfo.getDataStore().getRole())) {
```
to make sure no NPE happens?

Comment (harikrishna-patnala) [reply server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java]: Done, thanks @DaanHoogland

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Review by rohityadavcloud — APPROVED** [review]
LGTM didn't test it though

File: server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java [hunk 2]
```diff
@@ -101,15 +101,21 @@ public void expungeTemporarySnapshot(boolean kvmSnapshotOnlyInPrimaryStorage, Sn
...
 
-        logger.debug(String.format("Expunging snapshot [%s] due to it is a temporary backup to create a volume from snapshot. It is occurring because the global setting [%s]"
-          + " has the value [%s].", snapInfo.getId(), SnapshotInfo.BackupSnapshotAfterTakingSnapshot.key(), backupSnapshotAfterTakingSnapshot));
-
-        try {
-            snapshotService.deleteSnapshot(snapInfo);
-        } catch (CloudRuntimeException ex) {
-            logger.warn(String.format("Unable to delete the temporary snapshot [%s] on secondary storage due to [%s]. We still will expunge the database reference, consider"
-              + " manually deleting the file [%s].", snapInfo.getId(), ex.getMessage(), snapInfo.getPath()), ex);
+        if (!DataStoreRole.Image.equals(snapInfo.getDataStore().getRole())) {
```
Comment (DaanHoogland) [inline server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java:104]: actually still not sure it won't be a potential NPE ; `snapInfo` is not null but `snapInfo.getDataStore()` still may be.

Comment (harikrishna-patnala) [reply server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java:104]: Datastore/Role cannot be null based on the existing method calls and it should be null for further steps as well. Do you want me to keep a check anyways ?

Comment (DaanHoogland) [reply server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java:104]: maybe you are right @harikrishna-patnala (probably actually) but in this public method a snapshotInfo object may be passed that is constructed with a null-datastore. I think an extra check does not hurt here.

**Comment by rohityadavcloud** [comment]
@blueorangutan test ol8 kvm-ol8

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@harikrishna-patnala Unable to create volume from the snapshot, Not sure if it's related to this or different issue.

```
2024-06-27 09:56:52,592 DEBUG [o.a.c.s.s.SnapshotObject] (API-Job-Executor-4:ctx-e580ce2e job-54 ctx-5af933d0) (logid:a539fca6) Failed to update state:com.cloud.utils.fsm.NoTransitionException: Unable to transition to a new state from Ready via CreateOnlyRequested
2024-06-27 09:56:52,593 DEBUG [o.a.c.s.s.SnapshotServiceImpl] (API-Job-Executor-4:ctx-e580ce2e job-54 ctx-5af933d0) (logid:a539fca6) Failed to copy snapshot
com.cloud.utils.exception.CloudRuntimeException: Failed to update state: com.cloud.utils.fsm.NoTransitionException: Unable to transition to a new state from Ready via CreateOnlyRequested
	at org.apache.cloudstack.storage.snapshot.SnapshotObject.processEvent(SnapshotObject.java:239)
	at org.apache.cloudstack.storage.snapshot.SnapshotServiceImpl.backupSnapshot(SnapshotServiceImpl.java:357)
	at org.apache.cloudstack.storage.snapshot.DefaultSnapshotStrategy.backupSnapshot(DefaultSnapshotStrategy.java:193)
	at org.apache.cloudstack.snapshot.SnapshotHelper.backupSnapshotToSecondaryStorageIfNotExists(SnapshotHelper.java:146)
	at org.apache.cloudstack.engine.orchestration.VolumeOrchestrator.createVolumeFromSnapshot(VolumeOrchestrator.java:565)

2024-06-27 09:56:52,598 DEBUG [o.a.c.s.SnapshotHelper] (API-Job-Executor-4:ctx-e580ce2e job-54 ctx-5af933d0) (logid:a539fca6) Expunge template for Snapshot [2] is called for primary storage role. Not expunging it, but we will still expunge the database reference of the snapshot for image storage role if any

2024-06-27 09:56:52,625 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-4:ctx-e580ce2e job-54) (logid:a539fca6) Unexpected exception while executing org.apache.cloudstack.api.command.admin.volume.CreateVolumeCmdByAdmin
com.cloud.utils.exception.CloudRuntimeException: Failed to create volume: be56269c-a6b7-4b80-b3f6-6e921eec79fe
	at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:1060)
	at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:241)
...
Caused by: com.cloud.utils.exception.CloudRuntimeException: Failed to copy snapshot:com.cloud.utils.exception.CloudRuntimeException: Failed to update state: com.cloud.utils.fsm.NoTransitionException: Unable to transition to a new state from Ready via CreateOnlyRequested
	at org.apache.cloudstack.storage.snapshot.SnapshotServiceImpl.backupSnapshot(SnapshotServiceImpl.java:382)
	at org.apache.cloudstack.storage.snapshot.DefaultSnapshotStrategy.backupSnapshot(DefaultSnapshotStrategy.java:193)
	at org.apache.cloudstack.snapshot.SnapshotHelper.backupSnapshotToSecondaryStorageIfNotExists(SnapshotHelper.java:146)

2024-06-27 09:56:52,627 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-4:ctx-e580ce2e job-54) (logid:a539fca6) Complete async job-54, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"Failed to create volume: be56269c-a6b7-4b80-b3f6-6e921eec79fe"}
```

**Comment by harikrishna-patnala** [comment]
@sureshanaparti thanks for testing.

But I could not reproduce the issue with a new environment

I've 2 snapshots from two disks one with backup and one without
image

I've created 2 volumes and templates from those two snapshots
image

I've also created 2 templates from those two snapshots
image

They are successful, can you please try again testing @sureshanaparti

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@harikrishna-patnala Unable to create template from the backed up snapshot, check whether this is related to the same issue or not.

```
2024-07-15 07:17:32,499 DEBUG [o.a.c.s.s.SnapshotObject] (API-Job-Executor-5:ctx-24db41c0 job-51 ctx-b79ad463) (logid:e9288858) Failed to update state:com.cloud.utils.fsm.NoTransitionException: Unable to transition to a new state from Ready via CreateOnlyRequested
2024-07-15 07:17:32,500 DEBUG [o.a.c.s.s.SnapshotServiceImpl] (API-Job-Executor-5:ctx-24db41c0 job-51 ctx-b79ad463) (logid:e9288858) Failed to copy snapshot
com.cloud.utils.exception.CloudRuntimeException: Failed to update state: com.cloud.utils.fsm.NoTransitionException: Unable to transition to a new state from Ready via CreateOnlyRequested
	at org.apache.cloudstack.storage.snapshot.SnapshotObject.processEvent(SnapshotObject.java:239)
	at org.apache.cloudstack.storage.snapshot.SnapshotServiceImpl.backupSnapshot(SnapshotServiceImpl.java:357)
	at org.apache.cloudstack.storage.snapshot.DefaultSnapshotStrategy.backupSnapshot(DefaultSnapshotStrategy.java:193)
	at org.apache.cloudstack.snapshot.SnapshotHelper.backupSnapshotToSecondaryStorageIfNotExists(SnapshotHelper.java:146)
	at com.cloud.template.TemplateManagerImpl.createPrivateTemplate(TemplateManagerImpl.java:1681)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:107)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy236.createPrivateTemplate(Unknown Source)
	at org.apache.cloudstack.api.command.user.template.CreateTemplateCmd.execute(CreateTemplateCmd.java:297)
	at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:172)
	at com.cloud.api.ApiAsyncJobDispatcher.runJob(ApiAsyncJobDispatcher.java:112)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:654)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:45)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:602)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
2024-07-15 07:17:32,506 DEBUG [c.c.r.ResourceLimitManagerImpl] (API-Job-Executor-5:ctx-24db41c0 job-51 ctx-b79ad463) (logid:e9288858) Updating resource Type = template count for Account = 2 Operation = decreasing Amount = 1
2024-07-15 07:17:32,508 DEBUG [c.c.r.ResourceLimitManagerImpl] (API-Job-Executor-5:ctx-24db41c0 job-51 ctx-b79ad463) (logid:e9288858) Updating resource Type = secondary_storage count for Account = 2 Operation = decreasing Amount = (8.00 GB) 8589934592
2024-07-15 07:17:32,512 DEBUG [o.a.c.s.SnapshotHelper] (API-Job-Executor-5:ctx-24db41c0 job-51 ctx-b79ad463) (logid:e9288858) Expunge template for Snapshot [1] is called for primary storage role. Not expunging it, but we will still expunge the database reference of the snapshot for image storage role if any
2024-07-15 07:17:32,524 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-5:ctx-24db41c0 job-51) (logid:e9288858) Unexpected exception while executing org.apache.cloudstack.api.command.admin.template.CreateTemplateCmdByAdmin
com.cloud.utils.exception.CloudRuntimeException: Failed to copy snapshot:com.cloud.utils.exception.CloudRuntimeException: Failed to update state: com.cloud.utils.fsm.NoTransitionException: Unable to transition to a new state from Ready via CreateOnlyRequested
	at org.apache.cloudstack.storage.snapshot.SnapshotServiceImpl.backupSnapshot(SnapshotServiceImpl.java:382)
	at org.apache.cloudstack.storage.snapshot.DefaultSnapshotStrategy.backupSnapshot(DefaultSnapshotStrategy.java:193)
	at org.apache.cloudstack.snapshot.SnapshotHelper.backupSnapshotToSecondaryStorageIfNotExists(SnapshotHelper.java:146)
	at com.cloud.template.TemplateManagerImpl.createPrivateTemplate(TemplateManagerImpl.java:1681)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:107)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy236.createPrivateTemplate(Unknown Source)
	at org.apache.cloudstack.api.command.user.template.CreateTemplateCmd.execute(CreateTemplateCmd.java:297)
	at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:172)
	at com.cloud.api.ApiAsyncJobDispatcher.runJob(ApiAsyncJobDispatcher.java:112)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:654)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:45)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:602)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
2024-07-15 07:17:32,526 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-5:ctx-24db41c0 job-51) (logid:e9288858) Complete async job-51, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"Failed to copy snapshot:com.cloud.utils.exception.CloudRuntimeException: Failed to update state: com.cloud.utils.fsm.NoTransitionException: Unable to transition to a new state from Ready via CreateOnlyRequested"}
```

**Comment by harikrishna-patnala** [comment]
checking @sureshanaparti , thanks

**Pull request merged**